### PR TITLE
[AMLII-434] Add support for zstd compression in serializers

### PR DESCRIPTION
**Pull request by carlosroman**
### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/config/setup/config.go [hunk 1]
```diff
@@ -447,6 +447,7 @@ func InitConfig(config pkgconfigmodel.Config) {
 	config.BindEnvAndSetDefault("serializer_max_series_points_per_payload", 10000)
 	config.BindEnvAndSetDefault("serializer_max_series_payload_size", 512000)
 	config.BindEnvAndSetDefault("serializer_max_series_uncompressed_payload_size", 5242880)
+	config.BindEnvAndSetDefault("serializer_compressor_kind", "zlib")
```
Comment (iksaif) [inline pkg/config/setup/config.go:450]: We should document the possible values somewhere ?

File: pkg/serializer/serializer.go [hunk 2]
```diff
@@ -60,32 +57,29 @@ var (
...
 
-	protobufExtraHeadersWithCompression = make(http.Header)
 	for k := range protobufExtraHeaders {
 		protobufExtraHeadersWithCompression.Set(k, protobufExtraHeaders.Get(k))
 	}
 
 	if compression.ContentEncoding != "" {
-		jsonExtraHeadersWithCompression.Set("Content-Encoding", compression.ContentEncoding)
-		protobufExtraHeadersWithCompression.Set("Content-Encoding", compression.ContentEncoding)
+		kind := config.Datadog.GetString("serializer_compressor_kind")
```
Comment (iksaif) [inline pkg/serializer/serializer.go:79]: we should validate that we support this compression mechanism (at startup maybe ?)

File: pkg/serializer/serializer.go [hunk 3]
```diff
@@ -426,11 +433,21 @@ func (s *Serializer) SendProcessesMetadata(data interface{}) error {
 	if err != nil {
 		return fmt.Errorf("could not serialize processes metadata payload: %s", err)
 	}
-	compressedPayload, err := compression.Compress(payload)
+	kind := config.Datadog.GetString("serializer_compressor_kind")
+	var compressedPayload []byte
+	switch kind {
+	case "zstd":
```
Comment (iksaif) [inline pkg/serializer/serializer.go:439]: we should use constants here to avoid typos

File: pkg/serializer/serializer.go [hunk 4]
```diff
@@ -15,6 +15,8 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/DataDog/zstd"
```
Comment (iksaif) [inline pkg/serializer/serializer.go:18]: Why do we use DataDog's zstd here ? (worth adding a code comment)

File: pkg/serializer/split/split_test.go [hunk 5]
```diff
@@ -83,7 +86,14 @@ func testSplitPayloadsSeries(t *testing.T, numPoints int, compress bool) {
 		localPayload := payload.GetContent()
 
 		if compress {
-			localPayload, err = compression.Decompress(localPayload)
+			kind := config.Datadog.GetString("serializer_compressor_kind")
```
Comment (iksaif) [inline pkg/serializer/split/split_test.go:89]: this code is repeated multiple time, we should probably move it to a function

File: pkg/serializer/split/split.go [hunk 6]
```diff
@@ -169,7 +172,13 @@ func serializeMarshaller(m marshaler.AbstractMarshaler, compress bool, marshalFc
 		return nil, nil, err
 	}
 	if compress {
-		compressedPayload, err = compression.Compress(payload)
+		kind := config.Datadog.GetString("serializer_compressor_kind")
```
Comment (iksaif) [inline pkg/serializer/split/split.go:175]: this code is written twice, worth putting in a function

File: pkg/serializer/serializer.go [hunk 7]
```diff
@@ -511,3 +528,10 @@ func makeOrchestratorPayloads(msg types.ProcessMessageBody, hostName, clusterID
 	payloads := []*[]byte{&body}
 	return transaction.NewBytesPayloadsWithoutMetaData(payloads), extraHeaders, nil
 }
+
+func getContentEncoding(compressorKind string) string {
+	if compressorKind == "zlib" {
```
Comment (iksaif) [inline pkg/serializer/serializer.go:533]: We should use constants here too

File: pkg/serializer/internal/stream/compressor.go [hunk 8]
```diff
@@ -13,16 +13,40 @@ import (
 	"compress/zlib"
 	"errors"
 	"expvar"
+	"io"
 
 	"github.com/DataDog/datadog-agent/pkg/telemetry"
 	"github.com/DataDog/datadog-agent/pkg/util/compression"
```
Comment (d33d33) [inline pkg/serializer/internal/stream/compressor.go:19]: compression use [zstd_0](https://github.com/DataDog/datadog-agent/blob/main/pkg/util/compression/zstd.go#L11)

**Comment by carlosroman** [comment]
Thank you @iksaif  and @d33d33 for your feedback. Will get @rayz to look at your suggestions when he finishes the final PR for this feature. We have had to go a different route with this now after some of the changes made while turning some packages into components. Current work happening in this branch [here](https://github.com/DataDog/datadog-agent/tree/raymond.zhao/zstd-compressor).

**Pull request closed without merging**